src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +74,236 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
+        private static void AddUnmanagedDllResolver() => ModifyUnmanagedDllResolver(loadContext => loadContext.ResolvingUnmanagedDll += HostFxrResolver);
 
-            // First check for the DOTNET_ROOT environment variable, as it's often there as with, for example, dotnet format.
-            if (IntPtr.Size == 4)
+        private static void RemoveUnmanagedDllResolver() => ModifyUnmanagedDllResolver(loadContext => loadContext.ResolvingUnmanagedDll -= HostFxrResolver);
+
+        private static void ModifyUnmanagedDllResolver(Action<AssemblyLoadContext> resolverAction)
+        {
+            // For Windows hostfxr is loaded in the process.
+            if (!IsWindows)
             {
-                // 32-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT(x86)", exeName);
+                var loadContext = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
+                if (loadContext != null)
+                {
+                    resolverAction(loadContext);
+                }
             }
-            else if (IntPtr.Size == 8)
+        }
+
+        private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
+        {
+            var hostFxrLibName = "libhostfxr";
+            var libExtention = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "dylib" : "so";
+
+            if (!hostFxrLibName.Equals(libraryName))
             {
-                // 64-bit architecture
-                dotnetPath ??= FindDotnetFromEnvironmentVariable("DOTNET_ROOT", exeName);
+                return IntPtr.Zero;
             }
 
-            if (dotnetPath is null)
+            var hostFxrRoot = Path.Combine(DotnetPath.Value, "host", "fxr");
+            if (Directory.Exists(hostFxrRoot))
             {
-                // We will generally find the dotnet exe on the path, but on linux, it is often just a 'dotnet' symlink (possibly even to more symlinks) that we have to resolve
-                // to the real dotnet executable.
-                // This will work as often as just invoking dotnet from the command line, but we can be more confident in finding a dotnet executable by following
-                // https://github.com/dotnet/designs/blob/main/accepted/2021/install-location-per-architecture.md
-                // This can be done using the nethost library. We didn't do this previously, so I did not implement this extension.
-                foreach (string dir in Environment.GetEnvironmentVariable("PATH").Split(Path.PathSeparator))
+                var versionParser = new SemanticVersionParser();
+                // Load hostfxr from the highest version, because it should be backward-compatible
+                var hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                            .Select(str => versionParser.TryParse(str, out var version) ? version : null)
+                            .Max();
```

Comment:
I think you can combine these with:
```suggestion
                            .Max(str => versionParser.TryParse(str, out var version) ? version : null);
```

Though you might need to replace null with max version?